File: app-service-comments/src/test/java/com/github/arhor/aws/graphql/federation/comments/infrastructure/router/MainRouterTest.java
package com.github.arhor.aws.graphql.federation.comments.infrastructure.router;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.test.web.servlet.MockMvc;

import java.net.URI;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@AutoConfigureMockMvc(addFilters = false)
@WebMvcTest(controllers = ConfigureMainRouter.class)
class MainRouterTest {

    @Autowired
    private MockMvc http;

    @Test
    void should_return_empty_response_with_NO_CONTENT_status_trying_to_get_favicon() throws Exception {
        // Given
        final var requestURI = URI.create("/favicon.ico");

        // When
        final var result = http.perform(get(requestURI));

        // Then
        result
            .andExpect(status().isNoContent())
            .andExpect(content().string(""));
    }
}

File: app-service-users/src/test/kotlin/com/github/arhor/aws/graphql/federation/users/infrastructure/router/MainRouterTest.kt
package com.github.arhor.aws.graphql.federation.users.infrastructure.router

import com.github.arhor.aws.graphql.federation.starter.security.CurrentUser
import com.github.arhor.aws.graphql.federation.starter.security.CurrentUserRequest
import com.github.arhor.aws.graphql.federation.users.service.UserService
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.verify
import org.hamcrest.Matchers.contains
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import java.net.URI
import java.util.UUID

@AutoConfigureMockMvc(addFilters = false)
@WebMvcTest(controllers = [MainRouter::class])
class MainRouterTest {

    @Autowired
    private lateinit var http: MockMvc

    @MockkBean
    private lateinit var userService: UserService

    @Test
    fun `should return empty response with NO_CONTENT status trying to get favicon`() {
        // Given
        val requestURI = URI.create("/favicon.ico")

        // When
        val result = http.get(uri = requestURI)

        // Then
        result.andExpect {
            status { isNoContent() }
            content { string("") }
        }
    }

    @Test
    fun `should handle user verification request`() {
        // Given
        val requestURI = URI.create("/api/users/verify")

        val expectedReq = CurrentUserRequest(username = "test-username", password = "test-password")
        val expectedRes = CurrentUser(id = UUID.randomUUID(), authorities = listOf("ROLE_TEST_1", "ROLE_TEST_2"))
        val expectedContentType = MediaType.APPLICATION_JSON

        every { userService.getUserByUsernameAndPassword(any()) } returns expectedRes

        // When
        val result = http.post(uri = requestURI) {
            content = """
                {
                    "username": "${expectedReq.username}",
                    "password": "${expectedReq.password}"
                }
            """.trimIndent()
            contentType = expectedContentType
        }

        // Then
        result.andExpect {
            status { isOk() }

            content { contentType(expectedContentType) }

            jsonPath("$.id") { value(expectedRes.id.toString()) }
            jsonPath("$.authorities") { value(contains(*expectedRes.authorities.toTypedArray())) }
        }

        verify(exactly = 1) { userService.getUserByUsernameAndPassword(expectedReq) }
    }
}
